adam/batch_propagation.py

```diff
@@ -0,0 +1,143 @@
+        if self._summary is None:
+            return []
+        return [[(float(n) * M2KM) for n in sv.split()]
+                for sv in self._summary.splitlines()]
+
+
+class SinglePropagation(AdamObject):
+    def __init__(self, propagation_params, opm_params):
+        AdamObject.__init__(self)
+        self._propagation_params = propagation_params
+        self._opm_params = opm_params
+        self._ephemeris = None
+        self._final_state_vector = None
+
+    def set_ephemeris(self, ephemeris):
+        self._ephemeris = ephemeris
+
+    def set_final_state_vector(self, final_state_vector):
+        self._final_state_vector = None if final_state_vector is None else [
+            float(n) * M2KM for n in final_state_vector.split()]
+
+    def get_propagation_params(self):
+        return self._propagation_params
+
+    def get_opm_params(self):
+        return self._opm_params
+
+    def get_ephemeris(self):
+        return self._ephemeris
+
+    def get_final_state_vector(self):
+        return self._final_state_vector
+
+
+class BatchPropagations(AdamObjects):
+    def __init__(self, rest):
+        AdamObjects.__init__(self, rest, 'BatchPropagation')
+
+    def __repr__(self):
+        return "BatchPropagations module"
+
+    def _build_batch_propagation_creation_data(
+            self, propagation_params, opm_params, project_uuid):
+        data = {
+            'description': propagation_params.get_description(),
+            'templatePropagationParameters': {
+                'start_time': propagation_params.get_start_time(),
+                'end_time': propagation_params.get_end_time(),
+                'propagator_uuid': propagation_params.get_propagator_uuid(),
+                'step_duration_sec': propagation_params.get_step_size(),
+                'opmFromString': opm_params.generate_opm(),
+            },
+            'project': project_uuid,
+        }
+
+        return data
+
+    def insert(self, batch_propagation, project_uuid):
+        data = self._build_batch_propagation_creation_data(
+            batch_propagation.get_propagation_params(),
+            batch_propagation.get_opm_params(),
+            project_uuid)
+        batch_propagation.set_uuid(AdamObjects._insert(self, data))
+
+    def update_with_results(self, batch_propagation):
+        uuid = batch_propagation.get_uuid()
+        response = AdamObjects._get_json(self, uuid)
+        if response is None:
+            raise RuntimeError("Could not retrieve results for " + uuid)
+
+        batch_propagation.set_summary(response.get('summary'))
+
+    def get(self, uuid):
+        response = AdamObjects._get_json(self, uuid)
+        if response is None:
+            return None
+
+        opmParams = OpmParams.fromJsonResponse(
+            response['templatePropagationParameters']['opm'])
+        propParams = PropagationParams.fromJsonResponse(
+            response['templatePropagationParameters'], response.get('description'))
```

Comment:
Some places use `response.get` and some `response[]`. Should this be consistent or is it to avoid KeyError or something?

Reply:
Yes exactly, basically everything in [] is guaranteed to be present, everything in .get() is optional. Added a few notes here and there.